exec/java-exec/src/main/java/org/apache/drill/exec/store/log/LogRecordReader.java

```diff
@@ -0,0 +1,623 @@
+    openFile();
+    setupProjection();
+    defineVectors();
+  }
+
+  private void setupPattern() {
+    try {
+      pattern = Pattern.compile(formatConfig.regex);
+      Matcher m = pattern.matcher("test");
+      capturingGroups = m.groupCount();
+    } catch (PatternSyntaxException e) {
+      throw UserException
+          .validationError(e)
+          .message("Failed to parse regex: \"%s\"", formatConfig.regex)
+          .build(logger);
+    }
+  }
+
+  private void setupProjection() {
+    if (isSkipQuery()) {
+      projectNone();
+    } else if (isStarQuery()) {
+      projectAll();
+    } else {
+      projectSubset();
+    }
+  }
+
+  private void projectNone() {
+    columns = new ColumnDefn[]{new VarCharDefn("dummy", -1)};
+  }
+
+  private void openFile() {
+    InputStream in;
+    try {
+      in = dfs.open(new Path(fileWork.getPath()));
+    } catch (Exception e) {
+      throw UserException
+          .dataReadError(e)
+          .message("Failed to open open input file: %s", fileWork.getPath())
+          .addContext("User name", userName)
+          .build(logger);
+    }
+    reader = new BufferedReader(new InputStreamReader(in, Charsets.UTF_8));
+  }
+
+  private void projectAll() {
+    List<String> fields = new ArrayList<>();
+    fields.addAll(formatConfig.getFields());
+    for (int i = fields.size(); i < capturingGroups; i++) {
+      fields.add("field_" + i);
+    }
+    columns = new ColumnDefn[capturingGroups];
+
+    for (int i = 0; i < capturingGroups; i++) {
+      columns[i] = makeColumn(fields.get(i), i);
+    }
+
+
+  }
+
+  private void projectSubset() {
+    Collection<SchemaPath> project = this.getColumns();
+    assert !project.isEmpty();
+    columns = new ColumnDefn[project.size()];
+
+    List<String> fields = formatConfig.getFields();
+    int colIndex = 0;
+
+
+
+    for (SchemaPath column : project) {
+      if (column.getAsNamePart().hasChild()) {
+        throw UserException
+            .validationError()
+            .message("The log format plugin supports only simple columns")
+            .addContext("Projected column", column.toString())
+            .build(logger);
+      }
+      String name = column.getAsNamePart().getName();
+
+      //Need this to retrieve unnamed fields
+      Pattern r = Pattern.compile("^field_(\\d+)$");
+
+      // Now create matcher object.
+      Matcher m = r.matcher(name);
+      int patternIndex = -3;
+
+      if (m.find()) {
+        //if no fields are defined in the configuration, then all the fields have names of 'field_n'
+        //Therfore n is the column index
+        patternIndex = Integer.parseInt(m.group(1));
+      } else {
+        for (int i = 0; i < fields.size(); i++) {
+          if (fields.get(i).equalsIgnoreCase(name)) {
+            patternIndex = i;
+            break;
+          } else if( name.equals("_raw")){
+            patternIndex = -1;
+            break;
+          } else if( name.equals("_unmatched_rows")){
```

Comment:
`_unmatched_rows` sounds like there is more than one. But, we operate only at the row level. So, maybe `_unmatched` is more consistent with `_raw`.

But, we have `_raw`, so no need to repeat the whole row for `_unmatched`.

Maybe we need a config option about whether to ignore unmatched rows. If so, they'll never show up in the output. If we include them, then we just need a Boolean to say if the value is matched or not. So, instead of `_unmatched`, we might include a `_match`. Then, to get all the information:

```
SELECT a, b, _raw, _match FROM ...
```

I always get the `_raw` row. The `_match` tells me if the `a`, `b` values are filled in.

Or, maybe to be frugal, we keep the `_unmatched` which is like `_raw`, but is filled in only for unmatched rows.

Still, seems weird to have a single query return both matched and unmatched rows. I'd think I'd want to scan for unmatched rows so I can fix them, then have all other queries worry only about matched rows, else consumer of the data has to handle the match/unmatch issue. How, for example, would a BI tool chart values that contain unmatched rows?

So, is there a way to select either matched or unmatched rows? Taking a step back, maybe explain these use cases somewhere.

Reply:
To see all the rows that match, you could simply do a:
```
SELECT _raw FROM dfs.`log`;
```
IF you want to see the unmatched rows with no nulls:
```
SELECT `_unmatched_rows` 
FROM <data>
WHERE _unmatched_rows IS NOT NULL
```

Depending on the `max_errors` setting, the format-plugin pretty much ignores unmatched rows.